keps/0007-dynamic-audit-configuration.md

```diff
@@ -0,0 +1,342 @@
+
+## Summary
+
+We want to allow the advanced auditing features to be dynamically configured. Following in the same vein as [Dynamic Admission Control](https://kubernetes.io/docs/admin/extensible-admission-controllers/) we would like to provide a means of configuring the auditing features post cluster provisioning.
+
+## Motivation
+
+The advanced auditing features are a powerful tool, yet difficult to configure. The configuration requires deep insight into the deployment mechanism of choice and often takes many iterations to configure properly requiring a restart of the apiserver each time. Moreover, the ability to install addon tools that configure and enhance audting is hindered by the overhead in configuration. Such tools frequently run on the cluster requiring future knowledge of how to reach them when the cluster is live. These tools could enhance the security and conformance of the cluster and its applications.
+
+### Goals
+- Provide an api and set of objects to configure the advanced auditing kube-apiserver configuration dynamically
+
+### Non-Goals
+- Provide a generic interface to configure all kube-apiserver flags
+
+## Proposal
+
+### Dynamic Configuration
+Dynamic configuration of the advanced auditing features will be provided through use of a shared informer. Two levels of configuration will be provided, a cluster scoped object and a namespace scoped object.
+
+#### Cluster Scoped Configuration
+A cluster scoped configuration object will be provided that applies to all events in the cluster.
+
+```golang
+// ClusterAuditConfiguration represents a cluster level audit configuration
+type ClusterAuditConfiguration struct {
+    metav1.TypeMeta
+
+    v1.ObjectMeta
+
+    // Policy is the current audit v1beta1 Policy object
+    Policy *Policy `json:"policy"`
+
+    // Backends to send events
+    Backends *Backends `json:"backends"`
+}
+
+// Backends holds the configuration for the various event backends
+type Backends struct {
+    // Logs holds the file backends
+    Logs []*LogBackend  `json:"logs,omitempty"`
+
+    // Webhooks holds the webhook backends
+    Webhooks []*WebhookBackend `json:"webhooks,omitempty"`
+}
+
+// LogBackend holds the configuration for a single log backend
+type LogBackend struct {
+    // Path of file
+    Path string  `json:"path"`
+
+    // MaxAge in days to keep log files
+    MaxAge *int    `json:"maxAge,omitempty"`
+
+    // MaxBackup is the number of backups to maintain
+    MaxBackup *int  `json:"maxBackup,omitempty"`
+
+    // MaxSize of file in mb
+    MaxSize *int   `json:"maxSize,omitempty"`
+
+    // TruncateOptions follows the truncate options 
+    // https://github.com/kubernetes/apiserver/blob/master/pkg/server/options/audit.go#L157
+    TruncateOptions *AuditTruncateOptions `json:truncateOptions,omitEmpty"`
+}
+
+type WebhookMode string
+
+const (
+  WebhookModeBatch WebhookMode = "batch"
+  WebhookModeBlocking WebhookMode = "blocking"
```

Comment:
Agree that blocking mode is dangerous.  How is this different from mutating admission controllers?  Is there a way to turn them off so that you can reconfigure them? 

Regardless, this can be removed from now and discussed in a different PR if we think we really need it.

Reply:
Dynamic admission controllers do have this danger, but there are a bunch of escape hatches (e.g. excluding the admission controllers namespace from the namespaces it admits). I suppose we could do something similar by excluding ClusterAuditConfiguration changes from the blocked requests, but that feels strange since we want EXTRA auditing guarantees around changes to ClusterAuditConfiguration